### PR TITLE
Kappa analysis table

### PR DESCRIPTION
**Pull request by JingJZ160**
Adding a annotation comparison table so users can compare and correct their results. 

For each label, we have:
- One table for all users
- One table for each pair of distinct users

Code has been locally tested. It's best we do a screenshare to go through the workflow then pasting multiple screenshots here.

### CODE REVIEW

File: ar/data.py [hunk 1]
```diff
@@ -775,13 +781,27 @@ def _exclude_unknowns_for_kappa_calculation(result_user1, result_user2):
...
+        lambda: str))
     for label, df in kappa_matrices.items():
-        kappa_html_tables[label] = df.to_html(classes='kappa_table',
-                                              float_format=float_formatter)
-    return kappa_html_tables
+        columns = list(df.columns)
+        index = list(df.index.values)
+        for i in range(len(columns)):
+            user1 = columns[i]
+            user2 = index[i]
```
Comment (edu-gp) [inline ar/data.py:792]: Check this... for i in .. for j in .. how was this working?

File: ar/data.py [hunk 2]
```diff
@@ -907,3 +927,41 @@ def export_labeled_examples(task_id, outfile=None):
         save_jsonl(outfile, final)
 
     return final
+
+
+def _construct_comparison_df(dbsession, label: str, users_to_compare: List):
+    res = dbsession.query(
+        distinct(ClassificationAnnotation.entity)).filter(
+        ClassificationAnnotation.label == label).order_by(
+        ClassificationAnnotation.entity.asc()).all()
```
Comment (edu-gp) [inline ar/data.py:936]: Remember to add a task: Sort by how contentious annotations are in a future PR

File: ar/data.py [hunk 3]
```diff
@@ -907,3 +927,41 @@ def export_labeled_examples(task_id, outfile=None):
...
+
+    comparison_df = pd.DataFrame({
+        user: [str(np.NaN)] * len(distinct_entities_under_label)
+        for user in users_to_compare
+    }, index=list(distinct_entities_under_label))
+
+    annotation_id_per_user_df = comparison_df.copy(deep=True)
+
+    for user in users_to_compare:
+        entity_value_id_pairs = dbsession. \
```
Comment (edu-gp) [inline ar/data.py:950]: entity_value_id_pairs -> entity_value_id_tuples

File: ar/data.py [hunk 4]
```diff
@@ -907,3 +927,41 @@ def export_labeled_examples(task_id, outfile=None):
...
+            query(ClassificationAnnotation.entity,
+                  ClassificationAnnotation.value,
+                  ClassificationAnnotation.id). \
+            filter(ClassificationAnnotation.label == label,
+                   User.username == user). \
+            join(User). \
+            order_by(ClassificationAnnotation.entity.asc()).all()
+        entities = [pair[0] for pair in entity_value_id_pairs]
```
Comment (edu-gp) [inline ar/data.py:958]: tuple

File: frontend/static/annotation_box.js [hunk 5]
```diff
@@ -186,7 +186,7 @@ class AnnotationBox extends React.Component {
 
             var yes_btn_icon = String.fromCharCode(10003); // check
             var no_btn_icon = String.fromCharCode(10007); // cross
-
+            console.log(anno['labels'][label]);
```
Comment (edu-gp) [inline frontend/static/annotation_box.js:189]: debug message

File: ar/data.py [hunk 6]
```diff
@@ -556,16 +558,20 @@ def compute_annotation_statistics_db(dbsession, label):
...
     )
 
     return {
         'total_annotations': total_num_of_annotations_done_by_users,
         'total_distinct_annotated_entities': total_distinct_annotated_entities,
         'n_annotations_per_value': num_of_annotations_per_value,
         'n_annotations_per_user': n_annotations_done_per_user_dict,
-        'kappa_table': kappa_table_per_label,
+        'kappa_table': kappa_matrices,
+        'kappa_analysis_link_dict': kappa_analysis_dict_per_label
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply ar/data.py:574]: Done.

File: ar/data.py (hunk 3 above)
Comment (JingJZ160) [reply ar/data.py:950]: Done.

File: ar/data.py (hunk 4 above)
Comment (JingJZ160) [reply ar/data.py:958]: Done.

File: frontend/static/annotation_box.js (hunk 5 above)
Comment (JingJZ160) [reply frontend/static/annotation_box.js:189]: Done.

File: ar/data.py [hunk 7]
```diff
@@ -556,16 +558,20 @@ def compute_annotation_statistics_db(dbsession, label):
     kappa_stats_raw_data = _construct_kappa_stats_raw_data(
         db.session, distinct_users, label)
 
-    kappa_table_per_label = _convert_html_tables(
-        kappa_matrices=_compute_kappa_matrix(kappa_stats_raw_data)
+    kappa_matrices = _compute_kappa_matrix(kappa_stats_raw_data)
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply ar/data.py:561]: Done.

File: ar/data.py (hunk 1 above)
Comment (JingJZ160) [reply ar/data.py:792]: Done.

File: ar/data.py (hunk 2 above)
Comment (JingJZ160) [reply ar/data.py:947]: Created.

**Comment by JingJZ160** [comment]
Sorting by "contentious level" is included in this PR as well. It's a very simple implementation. Let me know if we want to go fancier.

**Pull request merged**